### PR TITLE
[Tests] Add more complex user switching tests

### PR DESCRIPTION
**Pull request by nan-li**
# Description
## One Line Summary
Add more complex user switching tests involving multiple calls to login / logout / user updates in a short span.

### Motivation

- Adding more testing, and prevent regressions in behavior. 
- We have seen bugs come up when changes have been made that changed expected behavior.

### Scope
Unit tests

# Testing
## Unit testing
Added new test class `SwitchUserTests`
1. `testIdentifyUserSuccessfully_thenLogin_sendsCorrectTags` - existing test migrated from other file
2. `testAnonUser_thenIdentifyUserWithConflict_sendsCorrectUpdatesAndFetchesUser` - Starts up with creating the anon user, then a login with conflict happens.
3. `testAnonUser_thenIdentifyUserWithConflict_thenLogout_sendsCorrectUpdatesWithNoFetch` - Starts up with creating the anon user, then a login with conflict happens, then a logout happens.
4. `testAnonUser_thenIdentifyUserWithConflict_thenLogin_sendsCorrectUpdatesAndFetchesUser` - Starts up with creating the anon user, then a login with conflict happens, then a login to another user happens.

## Manual testing
None, app runs.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [ ] Public API changes

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [x] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-iOS-SDK/1430)

### CODE REVIEW

File: iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift [hunk 1]
```diff
@@ -1,13 +1,62 @@
...
+     This method goes one level deep into dictionaries
+     */
+    private func contains(key: String, value: Any) -> Bool {
         guard let dictVal = self[key] else {
             return false
         }
-
-        return equals(dictVal, value)
+        if let value = value as? [String: Any],
+        let dictVal = dictVal as? NSDictionary {
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift:10]: Nit: This `if let` formating seems very odd, is this meet standard Swift linting?

File: iOS_SDK/OneSignalSDK/OneSignalUserTests/SwitchUserTests.swift [hunk 2]
```diff
@@ -0,0 +1,463 @@
+import XCTest
+import OneSignalCore
+import OneSignalCoreMocks
+import OneSignalUserMocks
+@testable import OneSignalUser
+
+final class SwitchUserTests: XCTestCase {
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/OneSignalUserTests/SwitchUserTests.swift:7]: Nit: Might be good to state these are more integration style tests.

**Review by jkasten2 — APPROVED** [review]

File: iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift (hunk 1 above)
Comment (nan-li) [reply iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift:10]: That's my bad, I did not align the `let`s.

Ideally this looks most correct to me:
```swift
if let value = value as? [String: Any],
   let dictVal = dictVal as? NSDictionary 
{
    return dictVal.contains(value)
} else {
    return equals(dictVal, value)
}
```

But running the swift auto-linter `swiftlint --fix`, the opening `{` does move to next line:
```swift
if let value = value as? [String: Any],
   let dictVal = dictVal as? NSDictionary {
    return dictVal.contains(value)
} else {
    return equals(dictVal, value)
}
```

Thoughts? There's a lot of these patterns and we can decide a better format we want to use.

Comment (nan-li) [reply iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift:10]: @jkasten2 That's my bad, I did not align the `let`s, and the swiftlint auto-linter did not fix it.

I think this looks most clear to me:
```swift
if let value = value as? [String: Any],
   let dictVal = dictVal as? NSDictionary 
{
    return dictVal.contains(value)
} else {
    return equals(dictVal, value)
}
```

But running the swift auto-linter `swiftlint --fix`, it moves the opening `{` up:
```swift
if let value = value as? [String: Any],
   let dictVal = dictVal as? NSDictionary {
    return dictVal.contains(value)
} else {
    return equals(dictVal, value)
}
```

Thoughts? There's a lot of these type of `if let` or `guard let` in the codebase, and we can decide a better format we want to use.

Comment (jkasten2) [reply iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift:10]: `swiftlint --fix` looks fine, I don't have any strong feels with that standard.

Original code really throw me off as; it was a combination of the where the 2nd `let` lined up with the `if` and not knowing you could use a `,` that allows multiple variables in the `if`.

File: iOS_SDK/OneSignalSDK/OneSignalUserTests/SwitchUserTests.swift (hunk 2 above)
Comment (nan-li) [reply iOS_SDK/OneSignalSDK/OneSignalUserTests/SwitchUserTests.swift]: Renamed class to `SwitchUserIntegrationTests`

File: iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift (hunk 1 above)
Comment (nan-li) [reply iOS_SDK/OneSignalSDK/OneSignalCoreMocks/Extensions/NSDictionary+UnitTests.swift]: Sounds good to me. Fixed up the original commit.

**Review by jkasten2 — APPROVED** [review]

**Pull request merged**